worker/codalabworker/aws_batch.py

```diff
@@ -0,0 +1,539 @@
+            jobDefinition=job_definition
+        )
+
+        job_id = response['jobId']
+
+        self.update_metadata(batch_job_id=job_id, submit_time=current_time())
+
+        return self.transition(Running)
+
+
+class JobFetcher(object):
+    def __init__(self, job_id, batch_client):
+        self._last_check_time = 0
+        self._check_frequency = 30  # seconds
+        self._job = None
+        self._job_id = job_id
+        self._batch_client = batch_client
+
+    def should_refresh(self):
+        now = current_time()
+        if now - self._last_check_time > self._check_frequency:
+            self._last_check_time = now
+            return True
+        return False
+
+    def fetch_job(self):
+        if self.should_refresh() or self._job is None:
+            response = self._batch_client.describe_jobs(jobs=[self._job_id])
+            self._job = response['jobs'][0]
+        return self._job
+
+
+class Running(AwsBatchRunState):
+    def __init__(self, *args, **kwargs):
+        super(Running, self).__init__(*args, **kwargs)
+        self._job_fetcher = JobFetcher(self._bundle['metadata']['batch_job_id'], self._batch_client)
+
+    """
+    Waiting for Batch to schedule the job
+    """
+    def update(self):
+        job = self._job_fetcher.fetch_job()
+        status = job['status']
+
+        now = current_time()
+        started = job.get('startedAt')
+        runtime = now - (started / 1000) if started else 0
+
+        run_status = 'Batch Status: %s' % status
+
+        self.update_metadata(run_status=run_status, last_updated=now, time=runtime)
+
+        if status in [BatchStatus.Succeeded, BatchStatus.Failed]:
+            return self.transition(Cleanup)
+
+        return self.transition_noop()
+
+
+class Cleanup(AwsBatchRunState):
+    def update(self):
+        self._batch_client.deregister_job_definition(
+            jobDefinition=self.metadata['batch_job_definition']
+        )
+
+        # TODO Cleanup the empty directories made by mounting the dependencies
+        return self.transition(Complete)
+
+
+class Complete(AwsBatchRunState):
+    def __init__(self, *args, **kwargs):
+        super(Complete, self).__init__(*args, **kwargs)
+        self._job_fetcher = JobFetcher(self._bundle['metadata']['batch_job_id'], self._batch_client)
+
+    def update(self):
+        job = self._job_fetcher.fetch_job()
+        status = job['status']
+        # The contain contains information about the most recent docker container used to run the job
+        container = job['container']
+
+        started = job.get('startedAt')
+        stopped = job.get('stoppedAt')
+
+        metadata_updates = {
+            'last_updated': current_time()
+        }
+
+        if started and stopped:
+            metadata_updates['time'] = (stopped - started) / 1000
+
+        if status == BatchStatus.Succeeded:
+            run_status = 'Succeeded'
+            finalize_message = {
+                'exitcode': container.get('exitCode'),
```

Comment:
Should this be `.get('exitCode', 0)` to match the Failed case? Or maybe they should both default to None?